### PR TITLE
Compact REPL printouts of `Associative`s should never be verbose

### LINKED ISSUE #15739: `Dict` REPL output lacks parentheses
```Julia
julia> a1 = Dict(('a'=>1) => ('b'=>2))
Dict('a'=>1=>'b'=>2)
```
The way this `Dict` is output lacks at least one set of parentheses. If we input this expression ...
```
julia> Dict('a'=>1=>'b'=>2)
Dict('a'=>1=>('b'=>2))
```
... we get a different result.

### PR DESCRIPTION
**Pull request by Tetralux**
Fix #15739. Again.

```
julia> d = Dict(1 => "hi"); show(d)
Dict{Any,Any}(Pair{Any,Any}(1,"hi"))
```

In this PR:

```
julia> d = Dict(1 => "hi"); show(d)
Dict(1=>"hi")
```

This also covers the previous cases from PR https://github.com/JuliaLang/julia/pull/15853, including `Pair`s as values and/or keys.

This was caused by `show(IO, Pair)` calling `show_default(IO, Pair)` if the type parameters
of the `Pair` did not match the actual types contained within. e.g: `Pair{Any,Any}(1, 2)`.

Also updated tests to cover this case, and be more robust.

cc @Keno @vtjnash

### CODE REVIEW

File: base/dict.jl [hunk 1]
```diff
@@ -75,12 +75,30 @@ function showdict{K,V}(io::IO, t::Associative{K,V}; compact = false)
...
             print(io, '(')
+
+            # @Duplicate of code in show(IO, Pair)
+            #
+            # Previously called `show` to print each Pair for the brackets
+            # that it inserted if either element of the Pair was itself a Pair.
+            # However, it calls `show_default` if the actual types of either element
+            # does not match the type parameters of the Pair.
+            # e.g: Pair{Any,Any}(1, 2)
+            print_pair_with_brackets = (io::IO, p::Pair) -> begin
```
Comment (Keno) [inline base/dict.jl]: I would rather not have this code duplicated, but rather factored out of show(io, Pair), such that both cases use the same code.

Comment (H-225) [inline base/dict.jl]: Agreed. I'll add some docs and comments while I'm at it.

Comment (vtjnash) [inline base/dict.jl]: I think it may be sufficient to declare that the `()=>()` style is always the compact form of showing a `Pair`

Comment (H-225) [inline base/dict.jl]: @vtjnash You'd end up with a non-trivial number of brackets, though.

`Dict{Any,Any}((1)=>(2=>0x33),(2)=>(3=>0x4444),(3)=>(25.0))`
`Dict{Any,Any}(1=>(2=>0x33),2=>(3=>0x4444),3=>25.0)`

Then again, that might be easier to read...

Comment (vtjnash) [inline base/dict.jl]: oh, I only meant `()=>()` as opposed to `show_default`

but then again, maybe it would be less ambiguous (since it might only needs `(a => b)` to resolve it)?

`Dict{Any,Any}((1 => (2 => 0x33)), (2 => (3 => 0x4444)), (3 => 25.0))`

Comment (H-225) [inline base/dict.jl]: @vtjnash Hmm... That is a fair bit easier to read - would it still qualify as a compact form though?... :)
It also might be better to put that into another PR? @Keno

**Comment by H-225** [comment]
My current intention is to submit a new PR which add the readability improvements suggested by @vtjnash, and leave this one as it is.

Unless you both think it would be fine to add it to this one?

**Comment by H-225** [comment]
Or indeed, have some other compelling reason. :)

**Comment by H-225** [comment]
I think it's ready, if it looks good. @Keno
